### DISCUSSION TITLE
EIP 1271 how to

### DISCUSSION THREAD

**Discussion by wiredmatt** [safe-global/safe-core-sdk#552] [Q&A] [answered]
Hey, our setup is an HSM signer (kp in this function) in a backend server. Off-chain signatures work fine with the safe transaction service, however we need on-chain signing to login to OpenSea, and other dApps that require it.

```typescript
async signMessage(
        message: SafeMessage['message'],
        kp: KeyPair,
        safeAddress: string,
        chainId: number = 5,
    ) {
        const userSafe = this.getSafe(safeAddress, kp, chainId);
        const resolvedUserSafe = await userSafe.safePromise;

        const result = await userSafe.signData(message); // off-chain sig

        const sig = await resolvedUserSafe
            .getEthAdapter()
            .getSignMessageLibContract({
                safeVersion: await resolvedUserSafe.getContractVersion(),
                chainId,
            })
            .signMessage(message as string); // on-chain sig

        console.log(sig);

        return result;
    }
```

However I seem to get: 

```
Error: Invalid SignMessageLib contract address
    at EthersAdapter.getSignMessageLibContract (/home/mateo/work/___/node_modules/@safe-global/safe-ethers-lib/src/EthersAdapter.ts:191:13)
```

And if I specify a custom contract address:

```
too many arguments: passed to contract (count=2, expectedCount=1, code=UNEXPECTED_ARGUMENT, version=contracts/5.7.0)
```

Any help on getting on-chain signing working is appreciated :), this is the last we're missing for a fully functional wallet.

Tried going through this but they never show how to actually sign: https://gallery.mirror.xyz/GzHNX1UvpNmBDQjQ0pQnE_wsdxoS0_yWocRGOssH4Xw

Same with this article: https://medium.com/mighty-bear-games/how-to-handle-gnosis-safe-connection-and-signature-e4b8e62cf14d

When is the signMessage function from the lib called? and how?

**Comment by wiredmatt** [comment]
https://github.com/safe-global/safe-core-sdk/pull/529# I just saw this draft, does this mean it is not currently possible to do this properly? should we wait or make a build out of this branch?

**Comment by mmv08** [comment]
Could you please explain why you need an onchain signature for opensea? It shouldn't matter to them which signature to use; they're just using EIP-1271 to validate the signature. Our interface works with opensea and off-chain signatures.

**Reply by wiredmatt** [reply]
If I do off-chain signing against the Service, I get back a successful response. To generate the signature and the message I used what's in the safe web repo (tryOffChainSigning function), then make a POST request to the messages endpoint, which returns a successful response. 

What comes next? Because if I return that to the frontend and they try to use the signature (preparedSignature), of course it won't match since that's the signature of the Owner of the safe, not of the Safe itself. We also tried using 0x but no luck either. 

We're just trying to log in to OpenSea through WC for a complete Demo. Literally the last piece we're missing.

Thank you for time

**Reply by wiredmatt** [reply, marked as answer]
**Update**

Due to our custom setup, it seems we were missing to assign the contract FallbackHandler, and we had no way of verifying the signature properly due to that. I'll close because I doubt this would help others.
